lib/src/rules/avoid_types_as_parameter_names.dart

```diff
@@ -75,11 +72,33 @@ class _Visitor extends SimpleAstVisitor<void> {
   }
 
   bool _isTypeName(AstNode scope, SimpleIdentifier node) {
+    // TODO (asashour): the below sections should be removed once
+    // https://github.com/dart-lang/sdk/issues/46753 is fixed
+    var parent = scope.parent;
+    if (parent is FunctionExpression &&
+        _isTypeParameter(parent.typeParameters, node)) {
+      return true;
+    }
+    var classDeclaration = parent?.thisOrAncestorOfType<ClassDeclaration>();
+    if (classDeclaration != null &&
+        _isTypeParameter(classDeclaration.typeParameters, node)) {
+      return true;
+    }
+    if (parent is GenericFunctionType &&
+        _isTypeParameter(parent.typeParameters, node)) {
+      return true;
+    }
     var result = context.resolveNameInScope(node.name, false, scope);
     if (result.isRequestedName) {
       var element = result.element;
       return element is ClassElement || element is TypeAliasElement;
```

Comment:
To support type parameters we also need `|| element is TypeParameterElement` here.